yjit/src/codegen.rs

```diff
@@ -4273,6 +4273,31 @@ fn jit_rb_str_concat(
     true
 }
 
+// Codegen for rb_ary_empty_p()
+fn jit_rb_ary_empty_p(
+    _jit: &mut JITState,
+    ctx: &mut Context,
+    asm: &mut Assembler,
+    _ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    _argc: i32,
+    _known_recv_class: *const VALUE,
+) -> bool {
+    let array_opnd = ctx.stack_pop(1);
+    let array_reg = asm.load(array_opnd);
+    let len_opnd = get_array_len(asm, array_reg);
+
+    asm.cmp(len_opnd, 0.into());
+    let bool_val = asm.csel_e(Qtrue.into(), Qfalse.into());
```

Comment:
```suggestion
    asm.test(len_opnd, len_opnd);
    let bool_val = asm.csel_z(Qtrue.into(), Qfalse.into());
```

Not sure if it's intentional for readability, but `TEST r/m64, r64` (`REX.W + 85 /r`) seems one byte shorter than `CMP r/m64, imm8` (`REX.W + 83 /7 ib`) on x86_64. The same thing applies to `jit_rb_str_empty`.

Reply:
I'll merge it but I think we also may want to do optimizations like that directly in the backend if we can.